File: testers/pg/java/src/java/adventureworks/userdefined/CustomCreditcardId.java
package adventureworks.userdefined;

import dev.typr.foundations.Bijection;
import dev.typr.foundations.PgText;
import dev.typr.foundations.PgType;
import dev.typr.foundations.PgTypes;

/** Type for the primary key of table `sales.creditcard` */
public record CustomCreditcardId(Integer value) {
  public static PgText<CustomCreditcardId> pgText =
      PgText.instance((v, sb) -> PgText.textInteger.unsafeEncode(v.value(), sb));
  public static PgType<CustomCreditcardId> pgType =
      PgTypes.int4.to(Bijection.of(CustomCreditcardId::new, v -> v.value()));
}

File: testers/pg/kotlin/src/kotlin/adventureworks/userdefined/CustomCreditcardId.kt
package adventureworks.userdefined

import dev.typr.foundations.PgText
import dev.typr.foundationskt.Bijection
import dev.typr.foundationskt.PgType
import dev.typr.foundationskt.PgTypes

/** Type for the primary key of table `sales.creditcard` */
@JvmInline
value class CustomCreditcardId(val value: Int) {
  companion object {
    @JvmStatic
    val pgText: PgText<CustomCreditcardId> = PgText.instance { v, sb -> PgText.textInteger.unsafeEncode(v.value, sb) }
    @JvmStatic
    val pgType: PgType<CustomCreditcardId> = PgTypes.int4.to(Bijection.of({ CustomCreditcardId(it) }, { it.value }))
    @JvmStatic
    val pgTypeArray: PgType<List<CustomCreditcardId>> = PgTypes.int4.array().to(Bijection.of(
      { list -> list.map { CustomCreditcardId(it) } },
      { list -> list.map { it.value } }
    ))
  }
}
